File: Runtime/TransitionPatchPrioritizationMap.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core

Class TransitionPatchPrioritizationMap
    Inherits STSimMapBase1(Of TransitionPatchPrioritization)

    Public Sub New(
        ByVal scenario As Scenario,
        ByVal items As TransitionPatchPrioritizationCollection)

        MyBase.New(scenario)

        For Each Item As TransitionPatchPrioritization In items
            Me.TryAddItem(Item)
        Next

    End Sub

    Public Function GetPatchPrioritization(
        ByVal transitionGroupId As Integer,
        ByVal iteration As Integer,
        ByVal timestep As Integer) As TransitionPatchPrioritization

        Return MyBase.GetItem(
            transitionGroupId,
            iteration,
            timestep)

    End Function

    Private Sub TryAddItem(ByVal item As TransitionPatchPrioritization)

        Try

            Me.AddItem(
                item.TransitionGroupId,
                item.Iteration,
                item.Timestep,
                item)

        Catch ex As STSimMapDuplicateItemException

            Dim template As String =
                "A duplicate transition patch prioritization was detected: More information:" & vbCrLf &
                "Transition Group={0}, Iteration={1}, Timestep={2}"

            ExceptionUtils.ThrowArgumentException(
                template,
                Me.GetTransitionGroupName(item.TransitionGroupId),
                STSimMapBase.FormatValue(item.Iteration),
                STSimMapBase.FormatValue(item.Timestep))

        End Try

    End Sub

End Class

File: src/Runtime/TransitionPatchPrioritizationMap.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2023 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using SyncroSim.Core;

namespace SyncroSim.STSim
{
    internal class TransitionPatchPrioritizationMap : STSimMapBase1<TransitionPatchPrioritization>
    {
        public TransitionPatchPrioritizationMap(Scenario scenario, TransitionPatchPrioritizationCollection items) : base(scenario)
        {
            foreach (TransitionPatchPrioritization Item in items)
            {
                this.TryAddItem(Item);
            }
        }

        public TransitionPatchPrioritization GetPatchPrioritization(int transitionGroupId, int iteration, int timestep)
        {
            return base.GetItem(transitionGroupId, iteration, timestep);
        }

        private void TryAddItem(TransitionPatchPrioritization item)
        {
            try
            {
                this.AddItem(item.TransitionGroupId, item.Iteration, item.Timestep, item);
            }
            catch (STSimMapDuplicateItemException)
            {
                string template = "A duplicate transition patch prioritization was detected: More information:" + Environment.NewLine + "Transition Group={0}, Iteration={1}, Timestep={2}";
                ExceptionUtils.ThrowArgumentException(template, this.GetTransitionGroupName(item.TransitionGroupId), STSimMapBase.FormatValue(item.Iteration), STSimMapBase.FormatValue(item.Timestep));
            }
        }
    }
}
